### PR TITLE
#20: Add HtKeepAliveResponse for support of Keep-Alive header

### PR DESCRIPTION
**Pull request by dgroup**
This PR for #20 contains:
 - Skeleton for HtKeepAliveResponse
 - 1 todo task for implementation of HtKeepAliveWire

### CODE REVIEW

**Comment by 0crat** [comment]
Job [#68](https://github.com/yegor256/cactoos-http/issues/68) is now in scope, role is `REV`

**Comment by dgroup** [comment]
@llorllale could you please review the PR ?

**Comment by 0crat** [comment]
@llorllale[/z](https://www.0crat.com/u/llorllale) everybody who has role `REV` is banned at [#68](https://github.com/yegor256/cactoos-http/issues/68); I won't be able to assign anyone automatically; consider assigning someone manually (as in [§19](http://www.zerocracy.com/policy.html#19)), or invite more people (as in [§51](http://www.zerocracy.com/policy.html#51)), or remove the job from the scope (as in [§14](http://www.zerocracy.com/policy.html#14))

**Comment by 0crat** [comment]
This pull request [#68](https://github.com/yegor256/cactoos-http/issues/68) is assigned to @victornoel[/z](https://www.0crat.com/u/victornoel), here is [why](https://www.0crat.com/footprint/C63314D6Z/65fa1fba-5861-487d-8f00-f22f1a8cc0f8); the budget is 15 minutes, see [§4](http://www.zerocracy.com/policy.html#4); please, read [§27](http://www.zerocracy.com/policy.html#27) and when you decide to accept the changes, inform @llorllale[/z](https://www.0crat.com/u/llorllale) (the architect) right in this ticket; if you decide that this PR should not be accepted ever, also inform the architect; this [blog post](http://www.yegor256.com/2015/02/09/serious-code-reviewer.html) will help you understand what is expected from a code reviewer; there will be no monetary reward for this job

**Comment by dgroup** [comment]
@victornoel 
>  we should introduce an Envelope 

I'm not sure that we should do it now because the current implementation might be changed during the implementation of `HtKeepAliveWire`.

I would like to come back to this question once `HtKeepAliveWire` is completed.

File: src/main/java/org/cactoos/http/HtKeepAliveResponse.java [hunk 1]
```diff
@@ -0,0 +1,143 @@
...
+     * Ctor.
+     * @param wire The wire
+     * @param req The request
+     * @param mtimeout The timeout for the connection usage in milliseconds
+     * @param rmax The maximum quantity of the requests within the connection
+     *  timeout
+     * @checkstyle ParameterNumberCheck (5 lines)
+     */
+    public HtKeepAliveResponse(
+        final Wire wire, final String req, final long mtimeout, final int rmax
```
[the remark this answers is not in the document]
Comment (dgroup) [reply src/main/java/org/cactoos/http/HtKeepAliveResponse.java]: @victornoel fixed, thank you.

File: src/main/java/org/cactoos/http/HtKeepAliveResponse.java [hunk 2]
```diff
@@ -0,0 +1,143 @@
...
+        final URI uri, final long mtimeout, final int rmax
+    ) {
+        this(
+            new HtWire(uri),
+            new UncheckedText(
+                new FormattedText(
+                    HtKeepAliveResponse.TEMPLATE,
+                    new UncheckedScalar<>(
+                        new Ternary<>(
+                            uri.getQuery() == null, "/", uri.getQuery()
```
[the remark this answers is not in the document]
Comment (dgroup) [reply src/main/java/org/cactoos/http/HtKeepAliveResponse.java]: @victornoel fixed, thanks a lot.

File: src/main/java/org/cactoos/http/HtKeepAliveResponse.java [hunk 3]
```diff
@@ -0,0 +1,143 @@
...
+     * @param mtimeout The timeout for the connection usage in milliseconds
+     * @param rmax The maximum quantity of the requests within the connection
+     *  timeout
+     * @todo #20:30min Replace the HtWire by HtKeepAliveWire which should
+     *  support the <em>Keep-Alive</em> header. It should has at least 1 `ctor`
+     *  with 3 parameters:
+     *  - `wire` which represents the HTTP connection;
+     *  - `timeout` in milliseconds;
+     *  - `rmax` the max quantity of within the timeout.
+     *  In case if the quantity of requests more than `rmax` but takes less than
```
[the remark this answers is not in the document]
Comment (dgroup) [reply src/main/java/org/cactoos/http/HtKeepAliveResponse.java]: @victornoel thank you for the corrections.

**Comment by victornoel** [comment]
@dgroup thanks :) let's ask @llorllale about https://github.com/yegor256/cactoos-http/pull/68#issuecomment-454441387

@llorllale if you are ok with @dgroup, then it's good to merge for me :)

File: src/test/java/org/cactoos/http/HtKeepAliveResponseTest.java [hunk 4]
```diff
@@ -0,0 +1,54 @@
...
+ * @since 0.1
+ * @checkstyle MagicNumberCheck (500 lines)
+ * @checkstyle JavadocMethodCheck (500 lines)
+ */
+public final class HtKeepAliveResponseTest {
+
+    @Test
+    public void worksFineByUri() throws IOException {
+        new FtRemote(new TkText("Hello, dude!")).exec(
+            home -> MatcherAssert.assertThat(
```
Comment (victornoel) [inline src/test/java/org/cactoos/http/HtKeepAliveResponseTest.java]: @dgroup also use cactoos-matchers `Assertion` instead of `assertThat`

**Comment by victornoel** [comment]
@llorllale scratch that, I added one more comment

File: src/test/java/org/cactoos/http/HtKeepAliveResponseTest.java (hunk 4 above)
Comment (dgroup) [reply src/test/java/org/cactoos/http/HtKeepAliveResponseTest.java]: Assertion isn't released yet.

**Comment by dgroup** [comment]
@victornoel thank you, let's wait the @llorllale .

File: src/test/java/org/cactoos/http/HtKeepAliveResponseTest.java (hunk 4 above)
Comment (victornoel) [reply src/test/java/org/cactoos/http/HtKeepAliveResponseTest.java]: @dgroup cactoos-matchers 0.13 has been released :)

**Comment by victornoel** [comment]
@dgroup cactoos-matchers 0.13 has been released, you can fix the comment above I believe

@llorllale an opinion on https://github.com/yegor256/cactoos-http/pull/68#issuecomment-454441387 ?

**Comment by llorllale** [comment]
@victornoel I think your proposal sounds right.

**Comment by victornoel** [comment]
@dgroup see ARC comment above

**Comment by dgroup** [comment]
@victornoel done.

**Comment by victornoel** [comment]
@dgroup you are missing the thing about the `Envelope` as explained in https://github.com/yegor256/cactoos-http/pull/68#pullrequestreview-191409546

**Comment by victornoel** [comment]
@dgroup ping? :)

**Comment by dgroup** [comment]
@victornoel done.

File: src/main/java/org/cactoos/http/HtResponseEnvelope.java [hunk 5]
```diff
@@ -0,0 +1,62 @@
...
+     * HTTP request.
+     */
+    private final Input request;
+
+    /**
+     * Ctor.
+     * @param wre The wire
+     * @param req The request
+     */
+    public HtResponseEnvelope(final Wire wre, final Input req) {
```
Comment (victornoel) [inline src/main/java/org/cactoos/http/HtResponseEnvelope.java]: @dgroup this should take an Input and delegate to it, the wire and request should stay in HtResponse!

File: src/main/java/org/cactoos/http/HtResponse.java [hunk 6]
```diff
@@ -41,17 +39,7 @@
...
-    /**
-     * The wire.
-     */
-    private final Wire wire;
-
-    /**
-     * HTTP request.
-     */
-    private final Input request;
+public final class HtResponse extends HtResponseEnvelope {
```
Comment (victornoel) [inline src/main/java/org/cactoos/http/HtResponse.java]: @dgroup this shouldn't have changed

File: src/main/java/org/cactoos/http/HtKeepAliveResponse.java [hunk 7]
```diff
@@ -123,13 +111,7 @@ public HtKeepAliveResponse(
...
         final Wire wire, final Input req, final long mtimeout, final int rmax
     ) {
-        this.wire = wire;
-        this.request = req;
-    }
-
-    @Override
-    public InputStream stream() throws IOException {
-        return this.wire.send(this.request).stream();
+        super(wire, req);
```
Comment (victornoel) [inline src/main/java/org/cactoos/http/HtKeepAliveResponse.java]: @dgroup this should be `super(new HtResponse(wire, req))`

**Comment by victornoel** [comment]
@dgroup see my comment, the envelope pattern is not exactly what you implemented

File: src/main/java/org/cactoos/http/HtResponseEnvelope.java (hunk 5 above)
Comment (dgroup) [reply src/main/java/org/cactoos/http/HtResponseEnvelope.java]: @victornoel its unclear for me. 
Please explain why and for what purpose.

Comment (victornoel) [reply src/main/java/org/cactoos/http/HtResponseEnvelope.java]: @dgroup see other examples of Envelopes, you can also check out https://github.com/victornoel/eo-envelopes that contains example as well as a link to yegor's blog on the subject. The envelope shouldn't have behaviour, just wrapping and delegatig an object of its own type.

**Comment by dgroup** [comment]
@victornoel done.

File: src/main/java/org/cactoos/http/HtKeepAliveResponse.java [hunk 8]
```diff
@@ -0,0 +1,117 @@
...
+     * @param wire The wire
+     * @param req The request
+     * @param mtimeout The timeout for the connection usage in milliseconds
+     * @param rmax The maximum quantity of the requests within the connection
+     *  timeout
+     * @checkstyle ParameterNumberCheck (5 lines)
+     */
+    @SuppressWarnings("PMD.UnusedFormalParameter")
+    public HtKeepAliveResponse(
+        final Wire wire, final Input req, final long mtimeout, final int rmax
```
Comment (victornoel) [inline src/main/java/org/cactoos/http/HtKeepAliveResponse.java]: @dgroup why is there unused parameters?

**Comment by victornoel** [comment]
@dgroup there is one last comment about unused parameters in some constructors...

**Comment by dgroup** [comment]
@victornoel done.

**Comment by victornoel** [comment]
@dgroup thanks, good work and thank you for your patience :)

@llorllale good to merge I think

**Comment by llorllale** [comment]
@rultor merge

**Comment by 0crat** [comment]
Code review was too long (29 days), architects (@llorllale) were penalized, see [§55](http://www.zerocracy.com/policy.html#55)

**Comment by 0crat** [comment]
The job [#68](https://github.com/yegor256/cactoos-http/issues/68) is now out of scope

**Comment by 0crat** [comment]
Order was finished: +15 point(s) just awarded to @victornoel[/z](https://www.0crat.com/u/victornoel)

**Comment by 0crat** [comment]
Job `gh:yegor256/cactoos-http#68` doesn't exist in WBS, can't create order

**Comment by 0crat** [comment]
Job `gh:yegor256/cactoos-http#68` doesn't exist in WBS, can't create order

**Comment by 0crat** [comment]
Job `gh:yegor256/cactoos-http#68` doesn't exist in WBS, can't create order

**Comment by 0crat** [comment]
Job `gh:yegor256/cactoos-http#68` doesn't exist in WBS, can't create order

**Comment by 0crat** [comment]
Job `gh:yegor256/cactoos-http#68` doesn't exist in WBS, can't create order

**Comment by 0crat** [comment]
Payment to `ARC` for a closed pull request, as in [§28](http://www.zerocracy.com/policy.html#28): +10 point(s) just awarded to @llorllale[/z](https://www.0crat.com/u/llorllale)

**Pull request merged**